src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -1009,28 +1015,39 @@ size_t ShenandoahHeapRegion::promote_humongous() {
   size_t spanned_regions = ShenandoahHeapRegion::required_regions(obj->size() * HeapWordSize);
   size_t index_limit = index() + spanned_regions;
 
-  log_debug(gc)("promoting humongous region " SIZE_FORMAT ", spanning " SIZE_FORMAT, index(), spanned_regions);
+  {
+    // We need to grab the heap lock in order to avoid a race when changing the affiliations of spanned_regions from
+    // young to old.
+    ShenandoahHeapLocker locker(heap->lock());
+    size_t available_old_regions = old_generation->adjusted_unaffiliated_regions();
+    if (spanned_regions <= available_old_regions) {
+      log_debug(gc)("promoting humongous region " SIZE_FORMAT ", spanning " SIZE_FORMAT, index(), spanned_regions);
+
+      // For this region and each humongous continuation region spanned by this humongous object, change
+      // affiliation to OLD_GENERATION and adjust the generation-use tallies.  The remnant of memory
+      // in the last humongous region that is not spanned by obj is currently not used.
+      for (size_t i = index(); i < index_limit; i++) {
+        ShenandoahHeapRegion* r = heap->get_region(i);
+        log_debug(gc)("promoting humongous region " SIZE_FORMAT ", from " PTR_FORMAT " to " PTR_FORMAT,
+                      r->index(), p2i(r->bottom()), p2i(r->top()));
+        // We mark the entire humongous object's range as dirty after loop terminates, so no need to dirty the range here
+        r->set_affiliation(OLD_GENERATION);
+        old_generation->increase_used(r->used());
+        young_generation->decrease_used(r->used());
+      }
+      // Then fall through to finish the promotion after releasing the heap lock.
+    } else {
+      return 0;
```

Comment:
I'm going to add a TODO comment here, so that we can think about changing this behavior.  I totally agree with your rationale.  Problem is that we have "assumptions" and "invariants" scattered throughout the existing implementation that need to be carefully reconsidered if we allow the rules to bend.  (For example: there are lots of size_t subtractions that may overflow to huge unmeaningful numbers, and if we run with ShenandoahVerify enabled, it will complain if the size of the generation exceeds it capacity.

Reply:
I also like your idea about just keeping primitive humongous objects in YOUNG.  That would allow their memory to be reclaimed much more quickly if and when they do become garbage.  OTOH, it may create an "unexpected surprise" to anyone who is carefully specifying the sizes of young-gen and old-gen.  Once we have auto-sizing of old- and young- fully working, this would be a good tradeoff to make.